purenix/src/Nix/Print.hs

```diff
@@ -82,74 +90,116 @@ style a _ Single = a
 style _ b Multi = b
 
 exprStyle :: ExprF Style -> Style
+exprStyle (Attrs _ [] []) = Single
+exprStyle (Attrs [] [(sty, _)] []) = sty
+exprStyle (Attrs [] [] [(_, sty)]) = sty
 exprStyle Attrs {} = Multi
 exprStyle Let {} = Multi
 exprStyle v = bool Single Multi $ elem Multi v
 
-exprPrec :: ExprF a -> Int
-exprPrec Var {} = 10
-exprPrec Num {} = 10
-exprPrec Path {} = 10
-exprPrec String {} = 10
-exprPrec Attrs {} = 10
-exprPrec List {} = 10
-exprPrec Sel {} = 9
-exprPrec App {} = 8
+newtype Precedence = Precedence Int deriving newtype (Num, Eq, Ord)
+
+data Associativity = AssocLeft | AssocRight | AssocNone
+  deriving (Eq, Show)
+
+exprAssoc :: ExprF a -> Associativity
+exprAssoc Sel {} = AssocLeft
+exprAssoc App {} = AssocLeft
+exprAssoc (Bin op _ _) = opAssoc op
+  where
+    opAssoc Update = AssocRight
+    opAssoc Equals = AssocNone
+    opAssoc And = AssocLeft
+exprAssoc _ = AssocNone
+
+-- | Expression precedence.
+-- See: https://nixos.org/manual/nix/stable/#sec-language-operators
+-- Opersators listed in the above table have a precedence of (15 - <listed precedence>)
+exprPrec :: ExprF a -> Precedence
+exprPrec Var {} = 15
+exprPrec Int {} = 15
+exprPrec Double {} = 15
+exprPrec String {} = 15
+exprPrec Attrs {} = 15
+exprPrec List {} = 15
+exprPrec Path {} = 15
+exprPrec Sel {} = 14
+exprPrec App {} = 13
+exprPrec Not {} = 8
 exprPrec (Bin op _ _) = opPrec op
   where
-    opPrec :: Op -> Int
-    opPrec Update = 5
-exprPrec Abs {} = 0
+    opPrec :: Op -> Precedence
+    opPrec Update = 6
+    opPrec Equals = 4
+    opPrec And = 3
+exprPrec Cond {} = 0
+exprPrec Lam {} = 0
 exprPrec Let {} = 0
 
-parens :: Bool -> Style -> Printer -> Printer
-parens False _ = id
-parens True sty = delimit sty '(' ')'
-
-parenthesize :: ExprF a -> Int -> Bool
-parenthesize Attrs {} = const False
-parenthesize Let {} = const False
-parenthesize List {} = (< 10)
-parenthesize e = (< exprPrec e)
-
-sepBy :: Printer -> [Printer] -> Printer
-sepBy sep = mconcat . intersperse sep
+parenthesize :: forall a b. (a -> Associativity) -> (a -> Precedence) -> (a -> b) -> (a -> b) -> ExprF a -> ExprF b
+parenthesize assoc prec no yes = go
+  where
+    below :: Precedence -> a -> b
+    below p a = if prec a < p then yes a else no a
+    bin :: (forall c. c -> c -> ExprF c) -> a -> a -> ExprF b
+    bin op l r = op (f l AssocLeft) (f r AssocRight)
+      where
+        f x a = case compare (prec x) (exprPrec $ op () ()) of
+          GT -> no x
+          EQ | assoc x == a -> no x
+          _ -> yes x
+    go :: ExprF a -> ExprF b
+    go (Attrs ih ihf f) = Attrs ih (ihf & traverse . _1 %~ yes) (f & traverse . _2 %~ no)
+    go (Let binds body) = Let (binds & traverse . _2 %~ no) (body & no)
+    go (List elems) = List (below 14 <$> elems)
+    go (App f x) = bin App f x
+    go (Bin op l r) = bin (Bin op) l r
+    go e = fmap (below (exprPrec e)) e
```

Comment:
I can only imagine the exact logic in this function must have been annoying to figure out.

It took me a while to work out exactly what was going on here.

Reply:
There's a lot of ways to do it, this was the best way I could think of of writing it in a way that isolated the paren logic. Although odds are that if I were to rewrite it again, it'd look different